src/sumcheck.rs

```diff
@@ -0,0 +1,322 @@
+        }
+
+        Ok(())
+    }
+}
+
+/// A quad describes an individual gate in the circuit, serialized according to struct Quad in [1]
+/// and the ad-hoc definition in [2] and [3].
+///
+/// [2] and [3] implement a `struct Quad` which in turn contains a vector of `Quad::corner`. Each
+/// `corner` value corresponds to the draft's `struct Quad`.
+///
+/// [1]: https://datatracker.ietf.org/doc/html/draft-google-cfrg-libzk-00#section-7.6
+/// [1]: https://github.com/google/longfellow-zk/blob/87474f308020535e57a778a82394a14106f8be5b/lib/proto/circuit.h
+/// [2]: https://github.com/google/longfellow-zk/blob/87474f308020535e57a778a82394a14106f8be5b/lib/sumcheck/quad.h
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Default)]
+pub struct Quad {
+    /// The position of ths gate in its layer.
+    gate_number: Size,
+    /// Index of the left-hand wire feeding into this gate.
+    left_wire: Size,
+    /// Index of the right-hand wire feeding into this gate.
+    right_wire: Size,
+    /// Index into the circuit's constant table.
+    const_table_index: Size,
+}
+
+impl Quad {
+    /// Encode this quad as deltas relative to the previous quad in the circuit.
+    // XXX extend [`crate::Codec`] to take an encoding parameter like prio::codec does so we can
+    // pass in prev_quad
+    fn encode(&self, prev_quad: Option<Quad>, bytes: &mut Vec<u8>) -> Result<(), anyhow::Error> {
+        let prev_quad = prev_quad.unwrap_or_default();
+
+        self.gate_number
+            .encode_delta(prev_quad.gate_number, bytes)?;
+        self.left_wire.encode_delta(prev_quad.left_wire, bytes)?;
+        self.right_wire.encode_delta(prev_quad.right_wire, bytes)?;
+        self.const_table_index.encode(bytes)?;
+
+        Ok(())
+    }
+
+    /// Encode this quad as deltas relative to the previous quad in the circuit.
+    fn decode(prev_quad: Option<Quad>, bytes: &mut Cursor<&[u8]>) -> Result<Self, anyhow::Error> {
+        let prev_quad = prev_quad.unwrap_or_default();
+
+        let gate_number = Size::decode_delta(prev_quad.gate_number, bytes)?;
+        let left_wire = Size::decode_delta(prev_quad.left_wire, bytes)?;
+        let right_wire = Size::decode_delta(prev_quad.right_wire, bytes)?;
+        let const_table_index = Size::decode(bytes)?;
+
+        Ok(Self {
+            gate_number,
+            left_wire,
+            right_wire,
+            const_table_index,
+        })
+    }
+}
+
+/// A serialized element of FieldP128.
+#[derive(Clone, Copy, Debug, Eq, PartialEq, PartialOrd, Ord, Default)]
+pub struct SerializedFieldP128(u128);
+
+impl Codec for SerializedFieldP128 {
+    fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), anyhow::Error> {
+        bytes
+            .write_u128::<LittleEndian>(self.0)
+            .context("failed to write u128")
+    }
+
+    fn decode(bytes: &mut Cursor<&[u8]>) -> Result<Self, anyhow::Error> {
+        Ok(Self(
+            bytes
+                .read_u128::<LittleEndian>()
```

Comment:
I'm guessing we need to validate that the field elements we pulled in are actually within the field here. Perhaps a TODO?